frontends/openssl-provider/common.c

```diff
@@ -0,0 +1,406 @@
+#include "common.h"
+
+#include "esdm_rpc_client.h"
+#include "helper.h"
+#include "math_helper.h"
+#include "openssl/core_names.h"
+#include "openssl/crypto.h"
+#include "openssl/evp.h"
+#include "openssl/params.h"
+#include "visibility.h"
+
+#include <errno.h>
+
+
+/************************************
+ * RAND specific provider functions *
+ ************************************/
+
+/* Context management */
+static OSSL_FUNC_rand_newctx_fn esdm_rand_newctx;
+static OSSL_FUNC_rand_freectx_fn esdm_rand_freectx;
+/* Random number generator functions: NIST */
+static OSSL_FUNC_rand_instantiate_fn esdm_rand_instantiate;
+static OSSL_FUNC_rand_uninstantiate_fn esdm_rand_uninstantiate;
+static OSSL_FUNC_rand_generate_fn esdm_rand_generate;
+static OSSL_FUNC_rand_reseed_fn esdm_rand_reseed;
+/* Random number generator functions: additional */
+static OSSL_FUNC_rand_nonce_fn esdm_rand_nonce;
+static OSSL_FUNC_rand_get_seed_fn esdm_rand_get_seed;
+static OSSL_FUNC_rand_clear_seed_fn esdm_rand_clear_seed;
+static OSSL_FUNC_rand_verify_zeroization_fn esdm_rand_verify_zeroization;
+/* Context Locking */
+static OSSL_FUNC_rand_enable_locking_fn esdm_rand_enable_locking;
+static OSSL_FUNC_rand_lock_fn esdm_rand_lock;
+static OSSL_FUNC_rand_unlock_fn esdm_rand_unlock;
+/* RAND parameter descriptors */
+static OSSL_FUNC_rand_gettable_ctx_params_fn esdm_rand_gettable_ctx_params;
+/* RAND parameters */
+static OSSL_FUNC_rand_get_ctx_params_fn esdm_rand_get_ctx_params;
+
+static void *esdm_rand_newctx(void *provctx, void *parent __unused,
+                              const OSSL_DISPATCH *parent_calls __unused)
+{
+  struct esdm_provider_ctx *cprov = provctx;
+  struct esdm_rand_ctx *rand = OPENSSL_secure_zalloc(sizeof(struct esdm_rand_ctx));
+
+  if (rand == NULL) {
+    goto err;
+  }
+
+  rand->core = cprov->core;
+  return rand;
+
+err:
+  OPENSSL_secure_clear_free(rand, sizeof(struct esdm_rand_ctx));
+  return NULL;
+}
+
+static void esdm_rand_freectx(void *ctx)
+{
+  struct esdm_rand_ctx *rand = ctx;
+
+  if (rand == NULL)
+    return;
+
+  CRYPTO_THREAD_lock_free(rand->lock);
```

Comment:
Sanity check: rand->lock may be NULL, can this OpenSSL handle NULL?

Reply:
At least in the version of OpenSSL I looked at (3.1.1):

```c
void CRYPTO_THREAD_lock_free(CRYPTO_RWLOCK *lock)
{
    if (lock == NULL)
        return;

# ifdef USE_RWLOCK
    pthread_rwlock_destroy(lock);
# else
    pthread_mutex_destroy(lock);
# endif
    OPENSSL_free(lock);

    return;
}
```